### PR TITLE
Example for modular tasnet

### PR DESCRIPTION
**Pull request by jensheit**
This example for tasnet allows to switch between the learnable encoder and decoder and the stft / istft. In future a ConvNet separator could be added to allow the user to switch between the DPRNN and ConvNet Separator.

The results achieved with this example are similar to the ones presented in [Demystifying TasNet: A Dissecting Approach](https://arxiv.org/abs/1911.08895)

### CODE REVIEW

File: padertorch/contrib/examples/tasnet/tas_coders.py [hunk 1]
```diff
@@ -127,3 +129,90 @@ def forward(self, w: torch.Tensor) -> torch.Tensor:
...
+    >>> import numpy as np
+    >>> stft_out = stft(\
+            mixture.numpy(), 256, 10, window_length=20, \
+            fading=None)
+    >>> stft_encoded = np.concatenate(\
+            [np.real(stft_out), np.imag(stft_out)], axis=-1\
+        ).transpose(0,1,3,2)
+    >>> np.testing.assert_allclose(stft_encoded, encoded, atol=1e-5)
+    """
+    def __init__(self, L: int = 20, N: int = 256, stride: int = None):
```
Comment (boeddeker) [inline padertorch/contrib/examples/tasnet/tas_coders.py:153]: I don't like `L` and `N`. Do you have a better name for these variables? I think they will now also appear in the config, so a longer name would be better.

File: padertorch/contrib/examples/tasnet/tas_coders.py [hunk 2]
```diff
@@ -127,3 +129,90 @@ def forward(self, w: torch.Tensor) -> torch.Tensor:
...
+        super().__init__()
+        self.L, self.N, self.stride = L, N, stride
+
+        if stride is None:
+            stride = L // 2
+        self.stft = STFT(size=N-2, shift=stride, window_length=L, fading=False,
+                         complex_representation='concat')
```
Comment (boeddeker) [inline padertorch/contrib/examples/tasnet/tas_coders.py:160]: Can you document, that you keep two uninformative values?
I think `N-2` is wrong. It may be something like: `(N - 1) // 2 * 2`

File: padertorch/contrib/examples/tasnet/tasnet.py [hunk 3]
```diff
@@ -226,3 +226,54 @@ def review(self, inputs, outputs):
...
+            mask: If `True`, use the output of the NN as a mask in tas domain
+                for separation. Otherwise, use the output directly as an
+                estimation for the separated signals.
+            output_nonlinearity: Nonlinearity applied to the output (right
+                before masking/decoding)
+            num_speakers: The number of speakers/output streams
+            additional_out_size: Size of the additional output. Has no effect if
+                set to 0.
+        """
+        super().__init__()
```
Comment (boeddeker) [inline padertorch/contrib/examples/tasnet/tasnet.py:257]: Does this work? I think `TasNet.__init__()` will complain.

File: padertorch/contrib/examples/tasnet/tas_coders.py (hunk 1 above)
Comment (jensheit) [reply padertorch/contrib/examples/tasnet/tas_coders.py:153]: I can change them to window_size and num_feats but in that case they should also be changed for the encoder and decoder

File: padertorch/contrib/examples/tasnet/tasnet.py [hunk 4]
```diff
@@ -226,3 +226,54 @@ def review(self, inputs, outputs):
 
     def flatten_parameters(self) -> None:
         self.dprnn.flatten_parameters()
+
+
+class ModularTasNet(TasNet):
```
Comment (boeddeker) [inline padertorch/contrib/examples/tasnet/tasnet.py:231]: Is the opposite inheritance not better? `TasNet` initialize encoder separator and decoder and forwards them to `ModularTasNet`.

For now I like to keep `TasNet` to make no breaking change. But in the long term, we may deprecate it.

File: padertorch/contrib/examples/tasnet/tasnet.py [hunk 5]
```diff
@@ -226,3 +226,54 @@ def review(self, inputs, outputs):
...
+    def __init__(
```
Comment (boeddeker) [inline padertorch/contrib/examples/tasnet/tasnet.py:232]: What do you think about a `finalize_dogmatic_config`?
In this way it is easier to switch between example configs.

File: padertorch/contrib/examples/tasnet/tas_coders.py [hunk 6]
```diff
@@ -127,3 +129,90 @@ def forward(self, w: torch.Tensor) -> torch.Tensor:
...
+
+    def forward(self, inputs, num_samples=None):
+        """
+        Args:
+            inputs: shape: [..., T], T is #samples
+            num_samples, list or tensor of #samples
```
Comment (thequilo) [inline padertorch/contrib/examples/tasnet/tas_coders.py:166]: Typo: missing `:`

File: padertorch/contrib/examples/tasnet/tas_coders.py (hunk 1 above)
Comment (boeddeker) [reply padertorch/contrib/examples/tasnet/tas_coders.py:153]: Since encoder and decoder is hidden for the user at the moment, we can change them too (and should do, to keep them consistent).

But as always: No shortcuts in the signature: `num_feats` -> `num_features`

File: padertorch/contrib/examples/tasnet/tas_coders.py [hunk 7]
```diff
@@ -127,3 +129,90 @@ def forward(self, w: torch.Tensor) -> torch.Tensor:
...
+            stft_signal: shape: [B, ..., N, Frames]
+            num_samples, list or tensor of #samples
+        Returns:
+            [B, ..., T]
+        """
+
+        stft_signal = rearrange(
+            stft_signal, '... fbins frames  -> ... frames fbins')
+
+        return self.stft.inverse(stft_signal)
```
Comment (thequilo) [inline padertorch/contrib/examples/tasnet/tas_coders.py:218]: Can you insert a newline at the end of the file?

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 4 above)
Comment (thequilo) [reply padertorch/contrib/examples/tasnet/tasnet.py:231]: I agree with this

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 5 above)
Comment (jensheit) [reply padertorch/contrib/examples/tasnet/tasnet.py:232]: yes, I though about that but I did not want to use it in an example since it is a little bit hard to follow

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 4 above)
Comment (jensheit) [reply padertorch/contrib/examples/tasnet/tasnet.py:231]: yes, the idea is to rework this after the workshop

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 5 above)
Comment (thequilo) [reply padertorch/contrib/examples/tasnet/tasnet.py:232]: You can also make sure that encoder and decoder configs are synchronized in there

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 3 above)
Comment (thequilo) [reply padertorch/contrib/examples/tasnet/tasnet.py:257]: It certainly will

File: padertorch/contrib/examples/tasnet/train_modular.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+import paderbox as pb
+import numpy as np
+
+from sacred.observers.file_storage import FileStorageObserver
+from lazy_dataset.database import JsonDatabase
+
+from padertorch.contrib.neumann.chunking import RandomChunkSingle
+from padertorch.contrib.ldrude.utils import get_new_folder
+
+nickname = "dprnn"
```
Comment (thequilo) [inline padertorch/contrib/examples/tasnet/train_modular.py:33]: `nickname="tasnet"`?

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 3 above)
Comment (jensheit) [reply padertorch/contrib/examples/tasnet/tasnet.py:257]: this works because tasnet has defaults for all inputs but yes it should be ```super(TasNet).__init__()```

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 4 above)
Comment (boeddeker) [reply padertorch/contrib/examples/tasnet/tasnet.py:231]: Do you mean deprecation or changing the inheritance order?

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 3 above)
Comment (boeddeker) [reply padertorch/contrib/examples/tasnet/tasnet.py:257]: I hate `super(TasNet).__init__()`, because it looks like python 2 and it is an indicator to rewrite some code.
As mentioned in another comment changing the inheritance order can solve this.

File: padertorch/contrib/examples/tasnet/tas_coders.py (hunk 1 above)
Comment (jensheit) [reply padertorch/contrib/examples/tasnet/tas_coders.py:153]: I used num_feats since it is used in tasnet and dprnn but I can change that as well

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 4 above)
Comment (jensheit) [reply padertorch/contrib/examples/tasnet/tasnet.py:231]: deprecation, I can change the inheritance now

File: padertorch/contrib/examples/tasnet/tas_coders.py (hunk 1 above)
Comment (boeddeker) [reply padertorch/contrib/examples/tasnet/tas_coders.py:153]: I didn't check the others. Use the name that you think is better.

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 4 above)
Comment (boeddeker) [reply padertorch/contrib/examples/tasnet/tasnet.py:231]: ok, thanks

File: padertorch/contrib/examples/tasnet/tas_coders.py (hunk 2 above)
Comment (jensheit) [reply padertorch/contrib/examples/tasnet/tas_coders.py:160]: N-2 is only wrong if we allow uneven stft sizes, which we do not support at the moment. But I can replace the size calculation and add an assert, that the feature_size is even.

File: padertorch/contrib/examples/tasnet/tasnet.py (hunk 3 above)
Comment (thequilo) [reply padertorch/contrib/examples/tasnet/tasnet.py]: `super(TasNet).__init__()` is wrong (as far as I understand). It should be `super(ModularTasNet, self).__init__()` or better `super().__init__()` and if you really want to skip the `__init__` of `TasNet`, you can write `super(TasNet, self).__init__()`. I generally don't think that skipping a constructor in the MRO is good coding style. And also it looks like Python2.

Comment (jensheit) [reply padertorch/contrib/examples/tasnet/tasnet.py]: I meant the later one but since we decided to change the inheritance order this problem is already addressed.
